src/vis_cpu/vis_cpu.py

```diff
@@ -20,66 +23,142 @@ def profile(fnc):
+        )
+
+    return beam_list, nbeam, beam_idx
+
+
+def _evaluate_beam_cpu(
+    beam_list,
+    tx,
+    ty,
+    polarized,
+    nbeam,
+    nax,
+    nfeed,
+    freq,
+    nsrcs_up,
+    complex_dtype,
+    spline_opts=None,
+):
+    A_s = np.zeros((nax, nfeed, nbeam, nsrcs_up), dtype=complex_dtype)
+
+    # Primary beam pattern using direct interpolation of UVBeam object
+    az, za = conversions.enu_to_az_za(enu_e=tx, enu_n=ty, orientation="uvbeam")
+    for i, bm in enumerate(beam_list):
+        kw = (
+            {
+                "reuse_spline": True,
+                "check_azza_domain": False,
+                "spline_opts": spline_opts,
+            }
+            if isinstance(bm, UVBeam)
+            else {}
         )
 
-    # Polarized beam
-    nax, nfeed, nbeam, bm_pix, _ = bm_cube.shape
+        interp_beam = bm.interp(
+            az_array=az,
+            za_array=za,
+            freq_array=np.atleast_1d(freq),
+            **kw,
+        )[0]
+
+        if polarized:
+            interp_beam = interp_beam[:, 0, :, 0, :]
+        else:
+            # Here we have already asserted that the beam is a power beam and
+            # has only one polarization, so we just evaluate that one.
+            interp_beam = np.sqrt(interp_beam[0, 0, 0, 0, :])
+
+        A_s[:, :, i] = interp_beam
+
+        # Check for invalid beam values
+        if np.any(np.isinf(A_s)) or np.any(np.isnan(A_s)):
+            raise ValueError("Beam interpolation resulted in an invalid value")
+
+    return A_s
+
+
+def _validate_inputs(precision, polarized, antpos, eq2tops, crd_eq, I_sky):
+    assert precision in {1, 2}
 
-    # x and y coordinates of beam
-    lm = conversions.bm_pix_to_lm(bm_pix)
+    # Specify number of polarizations (axes/feeds)
+    if polarized:
+        nax = nfeed = 2
+    else:
+        nax = nfeed = 1
 
-    # Construct splines for each polarization (pol. vector axis + feed) and
-    # antenna. The `splines` list has shape (Naxes, Nfeeds, Nants).
-    splines = []
-    for p1 in range(nax):
-        spl_axes = []
-        for p2 in range(nfeed):
-            spl_feeds = []
+    nant, ncrd = antpos.shape
+    assert ncrd == 3, "antpos must have shape (NANTS, 3)."
+    ntimes, ncrd1, ncrd2 = eq2tops.shape
+    assert ncrd1 == 3 and ncrd2 == 3, "eq2tops must have shape (NTIMES, 3, 3)."
+    ncrd, nsrcs = crd_eq.shape
+    assert ncrd == 3, "crd_eq must have shape (3, NSRCS)."
+    assert (
+        I_sky.ndim == 1 and I_sky.shape[0] == nsrcs
+    ), "I_sky must have shape (NSRCS,)."
 
-            # Loop over beams/antennas
-            for i in range(nbeam):
-                # Linear interpolation of primary beam pattern.
-                spl = RectBivariateSpline(lm, lm, bm_cube[p1, p2, i], kx=1, ky=1)
-                spl_feeds.append(spl)
-            spl_axes.append(spl_feeds)
-        splines.append(spl_axes)
-    return splines
+    return nax, nfeed, nant, ntimes
 
 
 @profile
 def vis_cpu(
```

Comment:
This is an API-breaking change, so the release notes would ideally mention how `bm_cube` users should update their code.

Reply:
Good point. Updated changelog.

Reply:
Good point -- noted in CHANGELOG